cmake/OpenCVModule.cmake

```diff
@@ -858,10 +858,14 @@ macro(_ocv_create_module)
 
   if((NOT DEFINED OPENCV_MODULE_TYPE AND BUILD_SHARED_LIBS)
       OR (DEFINED OPENCV_MODULE_TYPE AND OPENCV_MODULE_TYPE STREQUAL SHARED))
-    set_target_properties(${the_module} PROPERTIES COMPILE_DEFINITIONS CVAPI_EXPORTS)
+    target_compile_definitions(${the_module} PRIVATE CVAPI_EXPORTS)
     set_target_properties(${the_module} PROPERTIES DEFINE_SYMBOL CVAPI_EXPORTS)
   endif()
 
+  if (CMAKE_COMPILER_IS_GNUCXX AND CMAKE_BUILD_TYPE STREQUAL "Debug")
+    target_compile_definitions(${the_module} PUBLIC _GLIBCXX_DEBUG)
+  endif()
```

Comment:
No, they are not compatible, debug and release STL variants are completly different. But, since the compile definition is added as `PUBLIC`, it will be propagated to all dependent targets. For example. [here](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64_no_opt/builds/6557) you can see that build of sample apps (steps 8-9) has passed.

Or did you mean that user will not be able to use newly compiled debug version of OpenCV in existing app without recompilation? This is true, though such configurations are unlikely.

What about adding a build option `ENABLE_GNU_STL_DEBUG` turned `ON` by default for Debug builds?